data/tutorials/ds_05_seq.md

```diff
@@ -0,0 +1,317 @@
+| Functions | Mandatory | | |
+| Lists   | Mandatory |   |   |
+| Options | Recommended |   |   |
+| Arrays  | Nice to have |   |   |
+
+## Introduction
+
+Sequences look a lot like lists. However from a pragmatic perspective, one
+should imagine they may be infinite. That's the key intuition to understanding
+and using sequences.
+
+One way to look at a value of type `'a Seq.t` is to consider it as an icicle, a
+frozen stream of data. To understand this analogy, consider how sequences are
+defined in the standard library:
+```ocaml
+type 'a node =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = unit -> 'a node
+```
+This is the mutually recursive definition of two types; `Seq.node` which is
+almost the same as `list`:
+```ocaml
+type 'a list =
+  | []
+  | (::) of 'a * 'a list
+```
+and `Seq.t` which is merely a type alias for `unit -> 'a Seq.node`. The whole
+point of this definition is the type of the second argument of `Seq.Cons`, which
+is a function returning a sequence while its `list` counterpart returns a list. Let's
+compare the constructors of `list` and `Seq.node`:
+1. Empty lists and sequences are defined the same way, a constructor without any
+   parameter: `Seq.Nil` and `[]`.
+1. Non-empty lists and sequences are both pairs whose former member is a piece
+   of data;
+1. but the latter member, in lists, is a `list` too, while in sequences, it is a
+   function returning a `Seq.node`.
+
+A value of type `Seq.t` is “frozen” because the data it contains isn't
+immediately available, a `unit` value has to be supplied to recover it, and
+that's “unfreezing”. However, unfreezing only gives access to the tip of the
+icicle, since the second argument of `Seq.Cons` is a function too.
+
+Having frozen-by-function tails explains why sequences may be considered
+potentially infinite. Until a `Seq.Nil` value has been found in the sequence,
+one can't say for sure if some will ever appear. The sequence could be a stream
+of incoming requests in a server, readings from an embedded sensor or system logs.
+All have unforeseeable termination and it is easier to consider them infinite.
+
+In OCaml, any value `a` of type `t` can be turned into a constant function by
+writing `fun _ -> a`, which has type `'a -> t`. When writing `fun () -> a`
+instead, we get a function of type `unit -> t`. Such a function is called a
+[_thunk_](https://en.wikipedia.org/wiki/Thunk). Using this terminology, sequence
+values are thunks. With the analogy used earlier, `a` is frozen in its thunk.
+
+Here is how to build seemingly infinite sequences of integers:
+```ocaml
+# let rec ints n : int Seq.t = fun () -> Seq.Cons (n, ints (n + 1))
+val ints : int -> int Seq.t = <fun>
+```
+The function `ints n` look as if building the infinite sequence
+$(n; n + 1; n + 2; n + 3;...)$. In reality, since there isn't an infinite
+amount of distinct values of type `int`, those sequences are not increasing,
+when reaching `max_int` the values will circle down to `min_int`. They are
```

Comment:
```suggestion
When reaching `max_int`, the values will circle down to `min_int`. They are
```
It's not clear whether "when reaching..." is a continuation from the sequences not increasing (as in the "those sequences don't increase when reaching `max_int`") or if it goes with the next part, as shown above. Basically, if we cut it into two sentences, does the first sentence end with `max_int` or with "sequences don't increase"?